File: bindings/zig/src/canonical.zig
//! Canonicalization and content-addressed identity (spec/identity.md).
//!
//! The identity procedure, ported from bindings/python/causalontology/canonical.py:
//!   1. take the object as JSON,
//!   2. keep only the identity-bearing fields for its kind (with "type" injected),
//!   3. serialize with the JSON Canonicalization Scheme (RFC 8785),
//!   4. hash with SHA-256 (std.crypto.hash.sha2.Sha256),
//!   5. identifier = scheme + ":" + lowercase hex digest.

const std = @import("std");
const jcs = @import("jcs.zig");

const Value = jcs.Value;
const ObjectMap = jcs.ObjectMap;
const Allocator = jcs.Allocator;

/// One kind's identity recipe: its name, its identifier scheme prefix, and
/// the ordered list of identity-bearing fields.
pub const KindSpec = struct {
    kind: []const u8,
    prefix: []const u8,
    fields: []const []const u8,
};

/// The twenty-one kinds and their identity-bearing fields (IDENTITY_FIELDS +
/// PREFIX). Whole-word re-mint (P7): the scheme IS the type value for every
/// kind, so prefix == kind throughout. 3.0.0 adds the cross_stratal_seam and
/// the conduit's realized_by; 4.0.0 adds the attitude, the
/// predicted_occurrence, and the prediction_error - all additive and
/// identity-preserving: a record that omits a new field keeps its earlier
/// identifier byte-for-byte, and the new kinds open new identity schemes that
/// disturb no existing record.
pub const kind_specs = [_]KindSpec{
    // ---- type tier ----
    .{ .kind = "occurrent", .prefix = "occurrent", .fields = &.{ "label", "category", "stratum" } },
    .{ .kind = "causal_relation_object", .prefix = "causal_relation_object", .fields = &.{ "causes", "effects", "mechanism", "temporal", "modality", "context", "refines", "skips" } },
    .{ .kind = "continuant", .prefix = "continuant", .fields = &.{ "label", "category" } },
    .{ .kind = "realizable", .prefix = "realizable", .fields = &.{ "kind", "bearer", "label" } },
    .{ .kind = "stratum", .prefix = "stratum", .fields = &.{ "label", "scheme", "ordinal", "unit", "governs" } },
    .{ .kind = "bridge", .prefix = "bridge", .fields = &.{ "coarse", "fine", "relation" } },
    .{ .kind = "cross_stratal_seam", .prefix = "cross_stratal_seam", .fields = &.{ "source", "target", "mechanism_status", "chain" } },
    .{ .kind = "port", .prefix = "port", .fields = &.{ "bearer", "label", "direction", "accepts", "realizable" } },
    .{ .kind = "conduit", .prefix = "conduit", .fields = &.{ "label", "from", "to", "carries", "transform", "realized_by" } },
    .{ .kind = "quality", .prefix = "quality", .fields = &.{ "label", "datatype", "unit", "stratum" } },
    // ---- token tier ----
    .{ .kind = "token_individual", .prefix = "token_individual", .fields = &.{ "instantiates", "designator", "part_of" } },
    .{ .kind = "token_occurrence", .prefix = "token_occurrence", .fields = &.{ "instantiates", "interval", "participants", "locus", "observer" } },
    .{ .kind = "state_assertion", .prefix = "state_assertion", .fields = &.{ "subject", "quality", "value", "interval" } },
    .{ .kind = "token_causal_claim", .prefix = "token_causal_claim", .fields = &.{ "causes", "effects", "covering_law", "actual_delay", "counterfactual" } },
    .{ .kind = "attitude", .prefix = "attitude", .fields = &.{ "holder", "attitude_type", "content" } },
    .{ .kind = "predicted_occurrence", .prefix = "predicted_occurrence", .fields = &.{ "instantiates", "interval", "predictor", "strength" } },
    .{ .kind = "prediction_error", .prefix = "prediction_error", .fields = &.{ "predicted", "observed", "discrepancy" } },
    // ---- provenance tier ----
    .{ .kind = "assertion", .prefix = "assertion", .fields = &.{ "about", "source", "evidence_type", "evidence", "strength", "confidence", "timestamp", "evidenced_by" } },
    .{ .kind = "enrichment", .prefix = "enrichment", .fields = &.{ "about", "field", "entry", "source", "timestamp" } },
    .{ .kind = "retraction", .prefix = "retraction", .fields = &.{ "retracts", "source", "timestamp" } },
    .{ .kind = "succession", .prefix = "succession", .fields = &.{ "predecessor", "successor", "timestamp" } },
};

/// The spec for a kind name, or null if unknown.
pub fn specOfKind(kind: []const u8) ?*const KindSpec {
    for (&kind_specs) |*s| {
        if (std.mem.eql(u8, s.kind, kind)) return s;
    }
    return null;
}

/// The kind whose identifier scheme is `prefix` (KIND_OF_PREFIX), or null.
pub fn kindOfPrefix(prefix: []const u8) ?[]const u8 {
    for (&kind_specs) |*s| {
        if (std.mem.eql(u8, s.prefix, prefix)) return s.kind;
    }
    return null;
}

/// Infer an object's kind from its type field, id prefix, or shape.
pub fn inferKind(obj: ObjectMap) ![]const u8 {
    if (jcs.getString(obj, "type")) |t| return t;
    if (jcs.getString(obj, "id")) |id| {
        if (std.mem.indexOfScalar(u8, id, ':')) |i| {
            if (kindOfPrefix(id[0..i])) |k| return k;
        }
    }
    if (obj.contains("coarse") and obj.contains("fine")) return "bridge";
    if (obj.contains("causes") and obj.contains("effects")) return "causal_relation_object";
    if (obj.contains("retracts")) return "retraction";
    if (obj.contains("predecessor") and obj.contains("successor")) return "succession";
    if (obj.contains("field") and obj.contains("entry")) return "enrichment";
    if (obj.contains("evidence_type") or (obj.contains("about") and obj.contains("confidence"))) return "assertion";
    if (obj.contains("kind") and obj.contains("bearer")) return "realizable";
    // occurrents and continuants share a shape; the caller must pass kind.
    return error.CannotInferKind;
}

pub const IdentityBearing = struct {
    kind: []const u8,
    ib: ObjectMap,
};

/// The identity-bearing subset of an object, with type always present.
pub fn identityBearing(a: Allocator, obj: ObjectMap, kind_opt: ?[]const u8) !IdentityBearing {
    const kind = kind_opt orelse try inferKind(obj);
    const spec = specOfKind(kind) orelse return error.UnknownKind;
    var out = ObjectMap.init(a);
    try out.put("type", .{ .string = spec.kind });
    for (spec.fields) |field| {
        if (obj.get(field)) |v| try out.put(field, v);
    }
    return .{ .kind = spec.kind, .ib = out };
}

/// The RFC 8785 identity-bearing bytes of an object.
pub fn canonicalize(a: Allocator, obj: ObjectMap, kind_opt: ?[]const u8) ![]u8 {
    const r = try identityBearing(a, obj, kind_opt);
    return jcs.jcs(a, .{ .object = r.ib });
}

/// The content-addressed identifier: scheme + ":" + SHA-256 hex.
pub fn identify(a: Allocator, obj: ObjectMap, kind_opt: ?[]const u8) ![]u8 {
    const r = try identityBearing(a, obj, kind_opt);
    const bytes = try jcs.jcs(a, .{ .object = r.ib });
    const spec = specOfKind(r.kind).?;
    return std.fmt.allocPrint(a, "{s}:{s}", .{ spec.prefix, try sha256Hex(a, bytes) });
}

/// Lowercase-hex SHA-256 of a byte string.
pub fn sha256Hex(a: Allocator, data: []const u8) ![]u8 {
    var digest: [32]u8 = undefined;
    std.crypto.hash.sha2.Sha256.hash(data, &digest, .{});
    return jcs.hexLower(a, &digest);
}

File: bindings/r/R/canonical.R
# causalontology-r -- canonical.R
#
# Canonicalization and content-addressed identity, ported from the
# reference bindings/python/causalontology/canonical.py.
#
# The identity procedure of spec/identity.md:
#   1. take the object as JSON,
#   2. keep only the identity-bearing fields for its kind ("type" injected),
#   3. serialize with the JSON Canonicalization Scheme (RFC 8785),
#   4. hash with SHA-256,
#   5. identifier = scheme + ":" + lowercase hex digest.
#
# 2.0.0 (Principle P7): every identifier scheme is a whole English word;
# the scheme, the type value, and the id prefix are one and the same
# string. Kinds are keyed here by that whole word throughout.

# The identity-bearing fields of each of the twenty-one kinds (3.0.0 adds the
# cross_stratal_seam; the conduit gains realized_by; 4.0.0 adds the attitude,
# the predicted_occurrence, and the prediction_error - all additive and
# identity-preserving, so a record that omits a new field keeps its earlier
# identifier byte-for-byte, and the new kinds open new identity schemes that
# disturb no existing record). "type" is always injected, so it is not listed.
# Order is irrelevant (JCS sorts).
co_identity_fields <- list(
  # ---- type tier ----
  occurrent  = c("label", "category", "stratum"),
  causal_relation_object = c("causes", "effects", "mechanism", "temporal",
                             "modality", "context", "refines", "skips"),
  continuant = c("label", "category"),
  realizable = c("kind", "bearer", "label"),
  stratum    = c("label", "scheme", "ordinal", "unit", "governs"),
  bridge     = c("coarse", "fine", "relation"),
  cross_stratal_seam = c("source", "target", "mechanism_status", "chain"),
  port       = c("bearer", "label", "direction", "accepts", "realizable"),
  conduit    = c("label", "from", "to", "carries", "transform", "realized_by"),
  quality    = c("label", "datatype", "unit", "stratum"),
  # ---- token tier ----
  token_individual   = c("instantiates", "designator", "part_of"),
  token_occurrence   = c("instantiates", "interval", "participants",
                         "locus", "observer"),
  state_assertion    = c("subject", "quality", "value", "interval"),
  token_causal_claim = c("causes", "effects", "covering_law",
                         "actual_delay", "counterfactual"),
  attitude             = c("holder", "attitude_type", "content"),
  predicted_occurrence = c("instantiates", "interval", "predictor", "strength"),
  prediction_error     = c("predicted", "observed", "discrepancy"),
  # ---- provenance tier ----
  assertion  = c("about", "source", "evidence_type", "evidence", "strength",
                 "confidence", "timestamp", "evidenced_by"),
  enrichment = c("about", "field", "entry", "source", "timestamp"),
  retraction = c("retracts", "source", "timestamp"),
  succession = c("predecessor", "successor", "timestamp")
)

# Whole-word re-mint (P7): the scheme IS the type value for every kind.
co_prefix <- stats::setNames(names(co_identity_fields), names(co_identity_fields))
co_kind_of_prefix <- stats::setNames(names(co_prefix), unname(co_prefix))

# Infer an object's kind from its type field, id prefix, or shape.
co_infer_kind <- function(obj) {
  if (co_has(obj, "type")) return(as.character(obj[["type"]])[[1]])
  oid <- co_get(obj, "id")
  if (co_is_str(oid) && grepl(":", oid, fixed = TRUE)) {
    pre <- strsplit(oid, ":", fixed = TRUE)[[1]][[1]]
    if (pre %in% names(co_kind_of_prefix)) return(co_kind_of_prefix[[pre]])
  }
  if (co_has(obj, "coarse") && co_has(obj, "fine")) return("bridge")
  if (co_has(obj, "causes") && co_has(obj, "effects")) return("causal_relation_object")
  if (co_has(obj, "retracts")) return("retraction")
  if (co_has(obj, "predecessor") && co_has(obj, "successor")) return("succession")
  if (co_has(obj, "field") && co_has(obj, "entry")) return("enrichment")
  if (co_has(obj, "evidence_type") ||
      (co_has(obj, "about") && co_has(obj, "confidence"))) return("assertion")
  if (co_has(obj, "kind") && co_has(obj, "bearer")) return("realizable")
  stop("cannot infer kind (occurrents and continuants share a shape); ",
       "pass kind explicitly")
}

# The identity-bearing subset of an object, with type always present.
# Returns list(kind = <string>, ib = <co_obj>).
co_identity_bearing <- function(obj, kind = NULL) {
  if (is.null(kind)) kind <- co_infer_kind(obj)
  if (!(kind %in% names(co_identity_fields))) stop("unknown kind: ", kind)
  ib <- co_obj(type = kind)
  for (field in co_identity_fields[[kind]]) {
    if (co_has(obj, field)) ib[[field]] <- obj[[field]]
  }
  list(kind = kind, ib = ib)
}

# The RFC 8785 identity-bearing bytes of an object (a raw vector).
co_canonicalize <- function(obj, kind = NULL) {
  r <- co_identity_bearing(obj, kind)
  co_utf8(co_jcs(r$ib))
}

# The content-addressed identifier: scheme + ":" + SHA-256 hex.
co_identify <- function(obj, kind = NULL) {
  r <- co_identity_bearing(obj, kind)
  digest <- co_sha256_hex(co_utf8(co_jcs(r$ib)))
  paste0(co_prefix[[r$kind]], ":", digest)
}
